### PR TITLE
fix(gateway): scope per-entity fault GET/DELETE by hosted-app FQNs

### LINKED ISSUE #395: [BUG] handle_get_fault returns cross-entity faults for synthetic components (scope leak)
# Bug report

### Steps to reproduce

1. Launch a gateway with at least two synthetic / runtime-discovered components (any deployment without a manifest declaring the components is enough).
2. Trigger a fault on an app hosted by component A (for example by injecting a fault from a sensor node).
3. Query the same fault code on a *different* synthetic component B that does **not** host that app:
   ```
   curl -s "localhost:8080/api/v1/components/$COMP_B_ID/faults/$FAULT_CODE"
   ```

### Expected behavior

`404 Resource Not Found` (or similar) - the fault was not reported by any app hosted by component B, so the request should not return a body containing it.

### Actual behavior

`200 OK` with the fault body that was reported by component A. Cross-entity scope leak: any client with read access to component B can pull faults that were reported by entirely different components, as long as the fault_code matches.

### Root cause

`FaultHandlers::handle_get_fault` (`src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp:631`) passes `entity_info.namespace_path` as `source_id` to `FaultManager::get_fault_with_env`. The match logic in `fault_manager.cpp:280` is:

```cpp
if (!source_id.empty()) {
  bool matches = false;
  for (const auto & src : result.fault.reporting_sources) {
    if (src.rfind(source_id, 0) == 0) { matches = true; break; }
  }
  if (!matches) {
    result.success = false;
    result.error_message = "Fault not found for source: " + source_id;
    ...
  }
}
```

Synthetic components have empty `fqn` *and* empty `namespace_path`. The guard `if (!source_id.empty())` short-circuits, the `reporting_sources` check is skipped entirely, and the manager returns the raw fault without any scope filter. The handler then forwards it to the client.

Manifest components with a non-empty `namespace_path` work correctly because the prefix match against `reporting_sources` still runs.

### Severity / class

Different class from #393 - that one is a *false negative* (synthetic components return empty results from `/logs` and `/bulk-data`). This one is a *false positive* / scope leak: a synthetic component returns faults that belong to other entities. Potentially a security-relevant boundary issue if any RBAC role allows `viewer` access to specific components but not others.

`handle_clear_fault` (DELETE per-entity) takes the same path and is presumably leaky in the same way - clearing a fault scoped to component B may delete a fault reported by component A. Worth verifying once the fix is in flight.

### Why the simple "iterate hosted apps" pattern from #393 does not work here

`FaultManager::get_fault_with_env(fault_code, source_id)` accepts a single `source_id` string and runs prefix-match against `fault.reporting_sources`. To fix this properly the manager (and likely the underlying `GetFault` ROS service contract in `ros2_medkit_msgs`) needs to accept a list of allowed source FQNs and match against any of them, with exact match instead of prefix match. That is a service-contract change with downstream impact on every generated client (Python MCP, Rust selfpatch CLI, etc.).

### Related

- Same root cause as #380 (string-matching on `reporting_sources` instead of explicit entity id) - see [comment on #380](https://github.com/selfpatch/ros2_medkit/issues/380#issuecomment-4336673964) for the cross-link. Whichever option lands on #380 should also unblock the proper fix here.
- Distinct from #393 (false-negative empty-result bug on `/logs` and `/bulk-data`, same synthetic-component root cause but different fix).

### Environment

- ros2_medkit version: main
- ROS 2 distro: any (Jazzy / Humble / Rolling)
- OS: any

### Additional information

Suggested approach (subject to design discussion on #380):

1. Extend `GetFault.srv` request with an optional `string[] allowed_sources` field (backward-compatible default empty = no scope filter).
2. Change `FaultManager::get_fault_with_env` signature to accept `std::vector<std::string> allowed_sources`.
3. In `handle_get_fault` and `handle_clear_fault`, build the list via `cache.get_apps_for_component(entity_id)` + `effective_fqn()` and pass exact-match FQNs.
4. Drop the `if (!source_id.empty())` guard - if the caller passes an empty list, treat that as the explicit "no scope filter" choice (only valid for global handlers, never for per-entity routes).

### PR DESCRIPTION
**Pull request by bburda**
## Summary

Per-entity fault routes (`GET` / `DELETE /{entity-path}/faults/{fault_code}`) leaked faults across entities. When the addressed entity had an empty `namespace_path` (host-derived or synthetic components, manifest components without a `namespace` field, Areas, Functions), the transport-level prefix filter short-circuited:

- `GET` returned faults reported by apps that lived in entirely different entities.
- `DELETE` cleared them too, because `ClearFault.srv` carries no scope argument and the handler issued the clear without checking ownership first.

Both handlers now resolve the addressed entity to the set of FQNs of every hosted app via a new `HandlerContext::resolve_entity_source_fqns` helper and apply a strict all-sources scope check: a fault is in scope iff **every** entry in its `reporting_sources` is owned by the entity. Per-fault routes return `404` for any fault that fails the check; per-entity collection responses omit it; per-entity `DELETE` skips it. `DELETE` additionally fetches the fault and verifies scope before issuing the clear so it can never clear a fault that does not belong to the entity, and uses the new `ClearFault.srv` `skip_correlation_auto_clear` request field to opt out of cascade-clearing correlated symptoms that may live outside the entity.

`GetFault.srv` is unchanged. `ClearFault.srv` gains the `skip_correlation_auto_clear` request field - this changes the service type hash and breaks out-of-tree callers that invoke `/fault_manager/clear_fault` directly (the `ros2_medkit_fault_manager` README documents that pattern). The in-tree gateway client and server are updated together. See CHANGELOG for the full client-visible behavior delta.

## Issue

- closes #395

## Type

- [x] Bug fix
- [ ] New feature or tests
- [x] Breaking change
- [ ] Documentation only

## Testing

TDD red-green verified locally:

- `test_faults_scope_isolation` (new integration test, `lidar_sensor` + `temp_sensor` in hybrid mode with `demo_nodes_manifest.yaml`):
  - With the handler reverted, `test_get_fault_returns_404_on_unrelated_component` gets `200` with the leaked lidar fault body and `test_clear_fault_returns_404_on_unrelated_component` gets `204` (the unrelated component cleared the lidar fault).
  - With the fix in place, all 4 cases pass (404 on the unrelated component, 200 on the owning component, fault still present after the rejected DELETE). The 404 cases additionally assert the response body carries none of the success-shape keys (`item`, `environment_data`, `reporting_sources`, `snapshots`) and does not echo the owning app's FQN, so a future regression that flips the status code without sanitizing the body still fails.
- 10 unit tests on `FaultHandlers::fault_in_source_scope` covering exact-match, strict path-child, prefix collision, multi-source all-in-scope, multi-source partially-in-scope (the DELETE escalation case), empty scope, empty / missing / malformed `reporting_sources`, and a root-FQN edge.
- 7 unit tests on `HandlerContext::resolve_entity_source_fqns` covering App / Component / Area (with nested subarea recursion) / Function (with component-host expansion) / Unknown plus empty-effective-fqn and missing-from-cache edges.
- 2 routing tests pinning that per-entity DELETE sets `skip_correlation_auto_clear=true` and the global DELETE leaves it `false`.
- Full gateway unit suite passes; existing fault integration tests (`test_faults_api`, `test_locking_faults`, `test_triggers_faults`, `test_scenario_fault_inspection`, `test_scenario_fault_lifecycle`) pass.
- `colcon test --packages-select ros2_medkit_gateway` linters (clang_format, cppcheck, flake8, lint_cmake, pep257, xmllint) and `clang-tidy` build green.

Reviewer steps:

1. `colcon build --packages-up-to ros2_medkit_integration_tests`
2. `colcon test --packages-select ros2_medkit_gateway ros2_medkit_integration_tests --ctest-args -R 'test_faults_scope_isolation|test_handler_context|test_fault_handlers|test_fault_manager_routing'`

## Checklist

- [x] Breaking changes are clearly described (and announced in docs / changelog if needed)
- [x] Tests were added or updated if needed
- [x] Docs were updated if behavior or public API changed

## Backwards compatibility note

Three client-visible behavior changes are documented in the CHANGELOG:

1. **Per-entity fault routes**: clients that previously received `200` / `204` for a fault outside an entity's scope now get `404`. The all-sources rule also rejects **mixed-source** faults that include any out-of-entity reporter - use the global `GET /api/v1/faults` to see those.
2. **`ClearFault.srv` service hash change**: adding the `skip_correlation_auto_clear` request field changes the service type hash. Out-of-tree callers that invoke `/fault_manager/clear_fault` directly must rebuild against the new `ros2_medkit_msgs`.
3. **Per-entity collection responses** no longer include the global `muted_count` / `cluster_count` / `muted_faults` / `clusters` correlation metadata; that remains on the global `GET /api/v1/faults` route.

### CODE REVIEW

File: src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp [hunk 1]
```diff
@@ -40,6 +40,32 @@ namespace handlers {
...
+    return false;
+  }
+  for (const auto & src : fault["reporting_sources"]) {
+    if (!src.is_string()) {
+      continue;
+    }
+    const auto src_str = src.get<std::string>();
+    for (const auto & prefix : source_prefixes) {
+      if (src_str.rfind(prefix, 0) == 0) {
+        return true;
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp]: Fixed in 37693373. The scope check now enforces a path boundary: a source matches a scope FQN iff it equals the FQN or the byte after the prefix is `/`. Sibling nodes like `/ns/node_extra` no longer pass for an entity owning `/ns/node`. Pinned by `FaultInSourceScopeTest.PrefixCollidingNameIsNotInScope` and `SubpathOfScopedFqnIsInScope`.

File: src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp [hunk 2]
```diff
@@ -672,17 +705,30 @@ void FaultHandlers::handle_get_fault(const httplib::Request & req, httplib::Resp
...
+    auto result = fault_mgr->get_fault_with_env(fault_code, "");
 
     if (result.success) {
       // Build SOVD-compliant response from the transport-supplied JSON shape.
       // The transport handed back `result.data = { "fault": {...}, "environment_data": {...} }`.
       const auto & fault_json = result.data.value("fault", json::object());
+
+      const auto & cache = ctx_.node()->get_thread_safe_cache();
+      auto source_fqns = HandlerContext::resolve_entity_source_fqns(cache, entity_info);
+      if (!fault_in_source_scope(fault_json, source_fqns)) {
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp]: Fixed in 37693373. `fault_in_source_scope` now requires ALL `reporting_sources` to be in scope. A fault with any out-of-scope source returns 404, so the response can never carry a mixed-source reporting list or environment data tied to nodes the caller cannot read. Pinned by `FaultInSourceScopeTest.MultiSourcePartiallyInScopeIsOutOfScope`.

File: src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp [hunk 3]
```diff
@@ -769,6 +815,37 @@ void FaultHandlers::handle_clear_fault(const httplib::Request & req, httplib::Re
...
+            res, 503, ERR_SERVICE_UNAVAILABLE, "Failed to clear fault",
+            {{"details", get_result.error_message}, {entity_info.id_field, entity_id}, {"fault_code", fault_code}});
+      }
+      return;
+    }
+
+    const auto & cache = ctx_.node()->get_thread_safe_cache();
+    auto source_fqns = HandlerContext::resolve_entity_source_fqns(cache, entity_info);
+    const auto & fault_json = get_result.data.value("fault", json::object());
+    if (!fault_in_source_scope(fault_json, source_fqns)) {
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp]: Fixed in 37693373. The DELETE path now uses the same all-sources-in-scope check before the clear, so a fault that any out-of-scope entity also reports cannot be cleared via this entity. Pinned by the same `FaultInSourceScopeTest.MultiSourcePartiallyInScopeIsOutOfScope`. Also applied to `handle_clear_all_faults` collection routes via `filter_faults_by_sources`, which shares the helper.

File: src/ros2_medkit_gateway/src/http/handlers/handler_context.cpp [hunk 4]
```diff
@@ -347,6 +347,59 @@ void HandlerContext::send_json(httplib::Response & res, const json & data) {
...
+        collect_app_fqn(cache, app_id, fqns);
+      }
+      break;
+    }
+    case EntityType::AREA: {
+      for (const auto & comp_id : cache.get_components_for_area(entity.id)) {
+        for (const auto & app_id : cache.get_apps_for_component(comp_id)) {
+          collect_app_fqn(cache, app_id, fqns);
+        }
+      }
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/handler_context.cpp]: Fixed in 37693373. `resolve_entity_source_fqns` AREA branch now does a BFS over `cache.get_subareas()`, so `/areas/powertrain/...` resolves to apps under nested areas like `engine`. Pinned by `ResolveEntitySourceFqnsTest.AreaWalksNestedSubareasRecursively`, which seeds powertrain -> engine with components on the subarea and asserts the parent area picks them up.

File: src/ros2_medkit_gateway/src/http/handlers/handler_context.cpp [hunk 5]
```diff
@@ -347,6 +347,59 @@ void HandlerContext::send_json(httplib::Response & res, const json & data) {
...
+      break;
+    }
+    case EntityType::FUNCTION: {
+      auto agg_configs = cache.get_entity_configurations(entity.id);
+      for (const auto & node : agg_configs.nodes) {
+        if (!node.node_fqn.empty()) {
+          fqns.insert(node.node_fqn);
+        }
+      }
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/handler_context.cpp]: Fixed in 37693373. `resolve_entity_source_fqns` now walks `Function.hosts` directly and dispatches per host kind: app hosts go through `collect_app_fqn`, component hosts go through `collect_component_app_fqns`. Pinned by `ResolveEntitySourceFqnsTest.FunctionHostingComponentExpandsToComponentApps`, which exercises a mixed component+app host list.

File: src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp]: Addressed in 37693373. `fault_in_source_scope` was promoted to a public static method on `FaultHandlers`, and `test_fault_handlers.cpp` now has 10 direct unit tests covering: exact match, strict path-child match, prefix collision (the boundary case you raised), multi-source all-in-scope, multi-source partially-in-scope (DELETE escalation), empty scope set, empty/missing/malformed `reporting_sources`, and a root-FQN edge. Full HTTP handler tests with mocked `FaultManager` / `GatewayNode` would be a larger scaffolding change - the integration test plus these unit tests cover the same scenarios end-to-end.

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/handlers/handler_context.hpp [hunk 6]
```diff
@@ -339,6 +340,34 @@ class HandlerContext {
...
+   *   `effective_fqn()` returns "" for those, so the scope filter treats them
+   *   as having no addressable node and any fault read becomes a 404)
+   * - Component: `effective_fqn()` of every hosted app via `get_apps_for_component()`
+   * - Area: `effective_fqn()` of every app in every component under the area
+   * - Function: non-empty `node_fqn` of every entry in the function's aggregation
+   *   configuration (entries with empty `node_fqn` are skipped)
+   * - Unknown: empty set
+   *
+   * An empty result means "no apps are in scope" and callers must treat any
+   * fault as out-of-scope (i.e., return 404), never as "no filter".
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/handlers/handler_context.hpp]: Fixed in 37693373. Doxygen now describes the empty-result semantics generically and notes per-route mapping: per-fault routes return 404, list routes return 200 with an empty `items` array, clear-all returns 204. The "no filter" prohibition stays.

File: src/ros2_medkit_integration_tests/test/features/test_faults_scope_isolation.test.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+
+class TestFaultsScopeIsolation(GatewayTestCase):
+    """Per-entity fault routes must not leak faults across components."""
+
+    MIN_EXPECTED_APPS = 2
+    REQUIRED_APPS = {'lidar-sensor'}
+
+    def setUp(self):
+        super().setUp()
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_integration_tests/test/features/test_faults_scope_isolation.test.py:85]: Fixed in 37693373. `setUp` now asserts `GET /components/{LIDAR_OWNER_COMPONENT,OTHER_COMPONENT}` returns 200 before any scope-leak assertion runs. If discovery fails to surface either component the test now fails with an explicit "precondition failed" message instead of producing a misleading green from an entity-not-found 404.

File: src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp [hunk 8]
```diff
@@ -569,20 +595,27 @@ void FaultHandlers::handle_list_faults(const httplib::Request & req, httplib::Re
...
+    // For Apps, scope by the app's effective FQN set. We can't reuse the
+    // transport-level prefix filter (`list_faults(namespace_path, ...)`)
+    // because an App with a wildcard `ros_binding.namespace_pattern` produces
+    // an empty effective_fqn, the transport silently disables the filter, and
+    // every fault in the system would be returned. Same bug class as #395 for
+    // Components - fix it the same way.
+    const auto & cache = ctx_.node()->get_thread_safe_cache();
+    auto app_fqns = HandlerContext::resolve_entity_source_fqns(cache, entity_info);
+    auto result = fault_mgr->list_faults("", filter.include_pending, filter.include_confirmed, filter.include_cleared,
+                                         filter.include_healed, include_muted, include_clusters);
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp]: Fixed in 3e867374. The App branch of `handle_list_faults` now hardcodes `include_muted=false` and `include_clusters=false` when calling `list_faults` and no longer emits any of `muted_count` / `cluster_count` / `muted_faults` / `clusters` on per-entity responses. The `include_muted` / `include_clusters` URL params are inert on per-entity routes (documented in CHANGELOG); the global `GET /api/v1/faults` route keeps them. This brings the App branch in line with Component/Area/Function which already omitted that metadata.

File: src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp [hunk 9]
```diff
@@ -769,6 +815,37 @@ void FaultHandlers::handle_clear_fault(const httplib::Request & req, httplib::Re
     }
 
     auto fault_mgr = ctx_.node()->get_fault_manager();
+
+    // Verify the fault is in this entity's scope BEFORE clearing. The
+    // underlying ClearFault.srv has no scope argument, so without this
+    // check a DELETE on one entity's route would clear faults owned by
+    // any other entity that happens to share the fault_code. See #395.
+    auto get_result = fault_mgr->get_fault_with_env(fault_code, "");
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp:757]: Fixed in 3e867374. `ClearFault.srv` gains a `skip_correlation_auto_clear` request flag. `fault_manager_node` honors it: when true, the correlation engine is NOT consulted, so no symptom fault codes are added to the auto-clear cascade. Per-entity DELETE handlers (`handle_clear_fault`, `handle_clear_all_faults`) set it to `true`; the global `DELETE /api/v1/faults` keeps it `false` so cluster-wide clearing still works there. Pinned by two new routing tests (`FaultManagerRoutingTest.ClearFaultDefaultsToCorrelationAutoClear` and `ClearFaultForwardsSkipCorrelationFlag`).

File: src/ros2_medkit_gateway/src/http/handlers/handler_context.cpp [hunk 10]
```diff
@@ -347,6 +347,93 @@ void HandlerContext::send_json(httplib::Response & res, const json & data) {
...
+      break;
+    case EntityType::AREA:
+      collect_area_app_fqns(cache, entity.id, fqns);
+      break;
+    case EntityType::FUNCTION:
+      collect_function_app_fqns(cache, entity.id, fqns);
+      break;
+    case EntityType::UNKNOWN:
+      break;
+  }
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/handler_context.cpp:433]: Fixed in 3e867374. Function and Area branches of `handle_list_faults` and `handle_clear_all_faults` now route through `HandlerContext::resolve_entity_source_fqns`, so subarea recursion and component-host expansion apply uniformly to both per-fault and collection routes. The Component branch is also consolidated via the helper for parity (no behavioral change there).

File: src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp [hunk 11]
```diff
@@ -672,17 +705,30 @@ void FaultHandlers::handle_get_fault(const httplib::Request & req, httplib::Resp
...
       // The transport handed back `result.data = { "fault": {...}, "environment_data": {...} }`.
       const auto & fault_json = result.data.value("fault", json::object());
+
+      const auto & cache = ctx_.node()->get_thread_safe_cache();
+      auto source_fqns = HandlerContext::resolve_entity_source_fqns(cache, entity_info);
+      if (!FaultHandlers::fault_in_source_scope(fault_json, source_fqns)) {
+        HandlerContext::send_error(res, 404, ERR_RESOURCE_NOT_FOUND, "Fault not found",
+                                   {{"details", "Fault is not reported by any app within this entity's scope"},
+                                    {entity_info.id_field, entity_id},
+                                    {"fault_code", fault_code}});
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp]: Fixed in 3e867374. Both 404 sites now describe the actual all-sources rule: "every reporting source must be one of the entity's owned apps, and a mixed-source fault that includes any out-of-entity reporter is rejected to prevent cross-entity disclosure".

File: src/ros2_medkit_gateway/CHANGELOG.rst [hunk 12]
```diff
@@ -5,8 +5,9 @@ Changelog for package ros2_medkit_gateway
...
...code}``, ``GET /api/v1/{entity-path}/faults``, and ``DELETE /api/v1/{entity-path}/faults`` previously fell back to a prefix match against the entity's ``namespace_path``; when that was empty (host-derived / synthetic components, manifest components without a ``namespace`` field, Areas, Functions, and Apps with a wildcard ``ros_binding.namespace_pattern``) the scope filter was silently disabled and the routes exposed - and on ``DELETE``, cleared - faults reported by apps that belonged to entirely different entities. All four handlers now resolve the addressed entity to its hosted-app FQN set (via the new ``HandlerContext::resolve_entity_source_fqns`` helper) and apply that set as a positive scope filter. Per-fault routes return ``404 Resource Not Found`` when the fault's ``reporting_sources`` does not intersect the set; collection routes return an empty ``items`` array. The underlying ``GetFault.srv`` / ``ClearFault.srv`` contracts are unchanged. Any client that previously received a ``200`` / ``204`` (single fault) or a populated ``items`` array for faults outside an entity's scope will now get ``404`` or an empty list (`#395 <https://github.com/selfpatch/ros2_medkit/issues/395>`_)
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/CHANGELOG.rst]: Fixed in 3e867374. The CHANGELOG entry now spells out the strict all-sources scope check (every entry in `reporting_sources` must be owned by the entity), explicitly calls out the mixed-source 404 case, documents that per-entity collection responses no longer include `muted_count` / `cluster_count` / `muted_faults` / `clusters`, and notes the new `ClearFault.srv` `skip_correlation_auto_clear` flag that per-entity DELETE uses to opt out of correlation cascade-clearing.

File: src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp [hunk 13]
```diff
@@ -166,6 +170,28 @@ json extract_primary_value(const std::string & message_type, const json & full_d
...
+    return false;
+  }
+  if (!fault.contains("reporting_sources") || !fault["reporting_sources"].is_array()) {
+    return false;
+  }
+  const auto & sources = fault["reporting_sources"];
+  if (sources.empty()) {
+    return false;
+  }
+  for (const auto & src : sources) {
```
Comment (mfaferek93) [inline src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp:184]: This loop makes the check all-match: a fault is rejected if ANY reporting_source is out-of-entity. Faults are keyed by fault_code alone (fault_storage.cpp:74) and sources from different entities merge onto one fault, so a shared fault_code (e.g. sensor_timeout) reported by two entities' apps now 404s on BOTH per-entity routes and is un-clearable per-entity - only global /faults can see or clear it. Leak-proof, but a real visibility regression for shared codes. Is that intended? If yes, worth steering operators to the global route in the API docs; if not, consider scope-filtering the returned reporting_sources rather than rejecting the whole fault.

File: src/ros2_medkit_msgs/srv/ClearFault.srv [hunk 14]
```diff
@@ -21,6 +21,14 @@
...
 # Request fields
 string fault_code        # The fault_code to clear
+
+# Disable auto-cleanup of correlated symptom faults. When false (default),
+# clearing a root-cause fault also clears every symptom the correlation engine
+# attributes to it (auto_clear_with_root rules). When true, only the requested
+# fault_code is cleared and `auto_cleared_codes` in the response is empty.
+# Set this from scoped per-entity DELETE routes so a viewer of one entity
+# cannot cascade-clear symptom faults reported by apps in other entities.
+bool skip_correlation_auto_clear
```
Comment (mfaferek93) [inline src/ros2_medkit_msgs/srv/ClearFault.srv:31]: Adding a request field changes the service type hash - a contract change for any out-of-tree caller of ros2_medkit_msgs/srv/ClearFault (the README documents direct `ros2 service call`), even though in-tree client and server rebuild together. The PR summary says "contracts unchanged" and the Breaking-change box is unchecked; worth correcting one of them, especially as this package ships via rosdistro.

File: src/ros2_medkit_integration_tests/test/features/test_faults_scope_isolation.test.py [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+    def test_get_fault_returns_404_on_unrelated_component(self):
+        """GET on a different component must not return another entity's fault.
+
+        @verifies REQ_INTEROP_013
+        """
+        response = requests.get(
+            f'{self.BASE_URL}/components/{OTHER_COMPONENT}/faults/{LIDAR_FAULT_CODE}',
+            timeout=10,
+        )
+        self.assertEqual(
```
Comment (mfaferek93) [inline src/ros2_medkit_integration_tests/test/features/test_faults_scope_isolation.test.py:137]: This guards the leak fix but only checks the 404 status, not that the leaked body is gone. A regression that returns 404 with the out-of-entity fault still in the body would pass here. Worth also asserting the response body doesn't carry that fault's details, since non-disclosure is the actual security property under test.

File: src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp (hunk 13 above)
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp:184]: Intended. Going the other direction (filtering `reporting_sources` instead of rejecting) would either silently drop entries from the response - which breaks the SOVD invariant that `reporting_sources` enumerates every reporter - or special-case DELETE (which is global per `fault_code` and cannot be made entity-scoped without a separate `clear_fault_for_source` service).

Documented in the API reference under Faults Endpoints in cbbd8b18: the new per-entity-scope note explains the all-sources rule and steers operators to the global `/api/v1/faults` route for shared codes. The CHANGELOG already had this but it was buried inside one long sentence.

File: src/ros2_medkit_msgs/srv/ClearFault.srv (hunk 14 above)
Comment (bburda) [reply src/ros2_medkit_msgs/srv/ClearFault.srv:31]: Correct - the PR description and the CHANGELOG both understated this. Fixed in cbbd8b18: PR description now checks the Breaking change box and drops the "contracts unchanged" claim; CHANGELOG has a separate Breaking Changes bullet calling out the type-hash change and the rebuild requirement for out-of-tree callers of `/fault_manager/clear_fault`.

File: src/ros2_medkit_integration_tests/test/features/test_faults_scope_isolation.test.py (hunk 15 above)
Comment (bburda) [reply src/ros2_medkit_integration_tests/test/features/test_faults_scope_isolation.test.py:137]: Fixed in cbbd8b18. Both 404 tests now call `assert_body_does_not_leak_fault`, which checks the response body carries none of the success-shape keys (`item`, `environment_data`, `reporting_sources`, `snapshots`) and does not echo the owning app's FQN (`/perception/lidar/lidar_sensor`). A regression that flips the status code to 404 without sanitizing the body now fails the test.

**Review by mfaferek93 — APPROVED** [review]
